### PR TITLE
add homework for method_missing

### PR DESCRIPTION
**Pull request by frameq**
@nonkor need review

### CODE REVIEW

File: lesson4/hometask_frameq.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class LazyHash < Hash
+  def method_missing(name)
+    key = name.to_s
+    super.method_missing(key) unless self.key?(key)
```
Comment (nonkor) [inline lesson4/hometask_frameq.rb]: The way how you calling `super` is incorrect. `#super` (_a method_) is not a `superclass` (_an object_).
You need to just call `super`. That's it.

Comment (nonkor) [reply lesson4/hometask_frameq.rb]: And what's happened if your hash would contain a _symbol_, not a _string_? I don't think that `self.key?(key)` would handle it.

File: lesson4/hometask_frameq.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    if self[key].class == Proc
```
Comment (nonkor) [inline lesson4/hometask_frameq.rb]: **Do not check classes in Ruby directly.**
Use `#respond_to?` whenever it's possible or `#is_a?` / `#kind_of?` if you **really** need to check class type, not allowed interface.

File: lesson4/hometask_frameq.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      self[key] = self[key].call
+      self[key]
+    end
+    self[key]
```
Comment (nonkor) [inline lesson4/hometask_frameq.rb:8]: duplication with line 7.

File: lesson4/hometask_frameq.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  end
+
+  def lazy?(key)
+    self[key.to_s].class == Proc ? true : false
```
Comment (nonkor) [inline lesson4/hometask_frameq.rb]: The same as above.
And **never** wrap such simple condition to return explicitly `true` or `false`. It's redundant.
You line `self[key.to_s].class == Proc` anyway will return  `true` or `false` why you try to wrap it?

**Review by nonkor — CHANGES_REQUESTED** [review]

File: lesson4/hometask_frameq.rb [hunk 5]
```diff
@@ -1,16 +1,15 @@
 class LazyHash < Hash
   def method_missing(name)
     key = name.to_s
```
Comment (frameq) [inline lesson4/hometask_frameq.rb:3]: Here convert $name to string. Other problems where fixed.

Comment (nonkor) [reply lesson4/hometask_frameq.rb:3]: It's not enough. Just try to add symbol key into a hash and call it as a method.
See,

``` ruby
a[:symbol_key] = 'foo'
p a.symbol_key #=> `method_missing': undefined method `symbol_key' for {:symbol_key=>1}:LazyHash (NoMethodError)
```

File: lesson4/hometask_frameq.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+class LazyHash < Hash
+  def method_missing(name)
+    key = name.to_s
+    super(name) unless self.key?(key)
```
Comment (nonkor) [inline lesson4/hometask_frameq.rb:4]: Better to call `super` without params to [automatically forwards](http://rubylearning.com/satishtalim/ruby_overriding_methods.html) the arguments to the method of superclass.

File: lesson4/hometask_frameq.rb (hunk 5 above)
Comment (frameq) [reply lesson4/hometask_frameq.rb:3]: @nonkor 
If we have such hash:

``` ruby
a[:symbol_key] = lambda {'i am symbol'}
a[symbol_key] = lambda {'i am string'} 
```

what should we do?

**Pull request closed without merging**